diff --git a/core/modules/migrate/src/Plugin/migrate/source/SourcePluginBase.php b/core/modules/migrate/src/Plugin/migrate/source/SourcePluginBase.php
index 8d69418262..856613aa12 100644
--- a/core/modules/migrate/src/Plugin/migrate/source/SourcePluginBase.php
+++ b/core/modules/migrate/src/Plugin/migrate/source/SourcePluginBase.php
@@ -450,13 +450,17 @@ public function count($refresh = FALSE) {
       return -1;
     }

+    if (!$this->cacheCounts) {
+      return $this->doCount();
+    }
+
     if (!isset($this->cacheKey)) {
       $this->cacheKey = hash('sha256', $this->getPluginId());
     }

     // If a refresh is requested, or we're not caching counts, ask the derived
     // class to get the count from the source.
-    if ($refresh || !$this->cacheCounts) {
+    if ($refresh) {
       $count = $this->doCount();
       $this->getCache()->set($this->cacheKey, $count);
     }
diff --git a/core/modules/migrate/src/Plugin/migrate/source/SqlBase.php b/core/modules/migrate/src/Plugin/migrate/source/SqlBase.php
index 74bfe479a2..aa15d573a5 100644
--- a/core/modules/migrate/src/Plugin/migrate/source/SqlBase.php
+++ b/core/modules/migrate/src/Plugin/migrate/source/SqlBase.php
@@ -385,7 +385,7 @@ protected function fetchNextBatch() {
   /**
    * {@inheritdoc}
    */
-  public function count($refresh = FALSE) {
+  public function doCount($refresh = FALSE) {
     return (int) $this->query()->countQuery()->execute()->fetchField();
   }
